filetracker/scripts/recover.py

```diff
@@ -0,0 +1,130 @@
+"""Script for recovering filetracker storage consistency after failures.
+
+This script iterates over all existing links, removing broken ones and
+recalculating blob reference count from scratch, overwriting existing
+values in DB.
+
+It also iterates over blobs and removes blobs that are not linked.
+
+WARNING: this script does not use or respect locks, so DO NOT run
+this while storage is being used by a filetracker server.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import sys
+
+import six
+
+from filetracker.scripts import progress_bar
+from filetracker.servers.storage import FileStorage
+
+# Value used for aligning printed action names
+_ACTION_LENGTH = 25
+
+
+def main():
+    parser = argparse.ArgumentParser(
+            description='Restores storage consistency after failures')
+    parser.add_argument('root', help='root directory of filetracker storage')
+    parser.add_argument('-s', '--silent', action='store_true',
+            help='if set, progress bar is not printed')
+
+    args = parser.parse_args()
+    root, silent = args.root, args.silent
+
+    check_storage_format(root)
+
+    # Create a FileStorage object to use the same db settings as usual
+    file_storage = FileStorage(root)
+    db = file_storage.db
+
+    links_widgets = [
+            ' [', progress_bar.Timer(format='Time: %(elapsed)s'), '] ',
+            ' Checking links '.ljust(_ACTION_LENGTH),
+            ' ', progress_bar.Counter(), ' ',
+            progress_bar.BouncingBar()
+    ]
+
+    processed_links = 0
+    broken_links = 0
+    blob_links = {}
+
+    with progress_bar.conditional(show=not silent,
+                                  widgets=links_widgets) as bar:
+        for cur_dir, _, files in os.walk(file_storage.links_dir):
+            for file_name in files:
+                link_path = os.path.join(cur_dir, file_name)
+
+                if not os.path.islink(link_path):
```

Comment:
I don' understand the logic here.
`islink` returns true even for links (broken ones too), so if it's not a link it's a normal file. Shouldn't that be `exists`?

Reply:
I see. That makes sense.